### PR TITLE
Add missing dependency to py-entrypoints

### LINKED ISSUE #4327: Cannot install py-entrypoints on an airgapped system
I'm working with the following setup: I have a frontend, which has access to the general internet but I'm trying to build on an airgapped cluster which has only access to a local file system.

In order to install stuff on the cluster, I'm doing a
```
spack fetch -dependencies py-entrypoints
```
to put the sources on the local file system and the doing (essentially a slurm ```srun```)
```
cluster spack install -v py-entrypoints
```
the latter fails with a failed fetch of ```configparser@3.5```

Introducing 
```
depends_on('py-configparser@3.5:', when='^python@2.7', type=('build', 'run'))
```
for ```py-entrypoints``` leads to
```
File "/wang/environment/software/jessie/spack/split_test/lib/spack/spack/spec.py", line 1316, in dag_hash
    self.to_node_dict(), default_flow_style=True, width=maxint)
  File "/wang/environment/software/jessie/spack/split_test/lib/spack/spack/spec.py", line 1357, in to_node_
dict
    ) for name, dspec in sorted(deps.items())
  File "/wang/environment/software/jessie/spack/split_test/lib/spack/spack/spec.py", line 1316, in dag_hash
    self.to_node_dict(), default_flow_style=True, width=maxint)
  File "/wang/environment/software/jessie/spack/split_test/lib/spack/spack/spec.py", line 1357, in to_node_
dict
    ) for name, dspec in sorted(deps.items())
  File "/wang/environment/software/jessie/spack/split_test/lib/spack/spack/spec.py", line 1316, in dag_hash
    self.to_node_dict(), default_flow_style=True, width=maxint)
  File "/wang/environment/software/jessie/spack/split_test/lib/spack/spack/spec.py", line 1331, in to_node_
dict
    d.update(self.versions.to_dict())
  File "/wang/environment/software/jessie/spack/split_test/lib/spack/spack/version.py", line 651, in to_dic
t
    ('version', str(self[0]))
  File "/wang/environment/software/jessie/spack/split_test/lib/spack/external/ordereddict_backport.py", lin
e 45, in __init__
    self.__update(*args, **kwds)
  File "/wang/environment/software/jessie/spack/split_test/lib/spack/external/ordereddict_backport.py", lin
e 172, in update
    self[key] = value
  File "/wang/environment/software/jessie/spack/split_test/lib/spack/external/ordereddict_backport.py", lin
e 55, in __setitem__
    dict_setitem(self, key, value)
RuntimeError: maximum recursion depth exceeded while calling a Python object
```

### PR DESCRIPTION
**Pull request by adamjstewart**
@healther Can you see if this fixes #4327?

### CODE REVIEW

File: var/spack/repos/builtin/packages/py-configparser/package.py [hunk 1]
```diff
@@ -29,16 +29,13 @@ class PyConfigparser(PythonPackage):
...
-    homepage = "https://pypi.python.org/pypi/configparser"
-    url      = "https://pypi.python.org/packages/source/c/configparser/configparser-3.5.0.tar.gz"
+    homepage = "https://docs.python.org/3/library/configparser.html"
+    url      = "https://pypi.io/packages/source/c/configparser/configparser-3.5.0.tar.gz"
 
-    version('3.5.0', 'cfdd915a5b7a6c09917a64a573140538',
-            url="https://pypi.python.org/packages/7c/69/c2ce7e91c89dc073eb1aa74c0621c3eefbffe8216b3f9af9d3885265c01c/configparser-3.5.0.tar.gz")
+    version('3.5.0', 'cfdd915a5b7a6c09917a64a573140538')
 
-    depends_on('python@2.6:2.8,3.4:')
```
Comment (healther) [inline var/spack/repos/builtin/packages/py-configparser/package.py]: why did you remove that?

**Comment by healther** [comment]
Yes it does, but I have a couple of questions if you would indulge me :)

File: var/spack/repos/builtin/packages/py-entrypoints/package.py [hunk 2]
```diff
@@ -25,15 +25,22 @@
...
 
-class PyEntrypoints(Package):
+class PyEntrypoints(PythonPackage):
     """Discover and load entry points from installed packages."""
 
     homepage = "https://pypi.python.org/pypi/entrypoints"
-    url      = "https://files.pythonhosted.org/packages/f8/ad/0e77a853c745a15981ab51fa9a0cb4eca7a7a007b4c1970106ee6ba01e0c/entrypoints-0.2.2-py2.py3-none-any.whl"
+    url      = "https://pypi.python.org/packages/f8/ad/0e77a853c745a15981ab51fa9a0cb4eca7a7a007b4c1970106ee6ba01e0c/entrypoints-0.2.2-py2.py3-none-any.whl"
+
+    import_modules = ['entrypoints']
```
Comment (healther) [inline var/spack/repos/builtin/packages/py-entrypoints/package.py]: what does this line do?

File: var/spack/repos/builtin/packages/py-configparser/package.py (hunk 1 above)
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-configparser/package.py]: I'm not aware of any such version restriction.

File: var/spack/repos/builtin/packages/py-entrypoints/package.py (hunk 2 above)
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-entrypoints/package.py]: See `spack edit -b python`. If you run `spack install --run-tests py-entrypoints`, it will try to import the modules from `import_modules` to see if they can be successfully imported. This catches a lot of bugs, especially for bigger packages like numpy.

File: var/spack/repos/builtin/packages/py-configparser/package.py (hunk 1 above)
Comment (healther) [inline var/spack/repos/builtin/packages/py-configparser/package.py]: From the docstring:
> This library brings the updated configparser from Python 3.5 to Python 2.6-3.5.
So it should be
```depends_on('python@:3.4.99')```
shouldn't it?

File: var/spack/repos/builtin/packages/py-entrypoints/package.py (hunk 2 above)
Comment (healther) [inline var/spack/repos/builtin/packages/py-entrypoints/package.py]: Ah I see, I wasn't working with the current ```LLNL/spack``` but with an older version, I guess time to rebase again

File: var/spack/repos/builtin/packages/py-configparser/package.py (hunk 1 above)
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-configparser/package.py]: Theoretically. I've seen a lot of bugs when packages use things like:
```python
depends_on('py-foo', when='^python@:2.6')
```
meaning that we have to either not support those dependencies (what we did previously, which is why it broke for you) or unconditionally add them for all versions. If I add:
```python
depends_on('python@:3.4.99')
```
it may be difficult for other packages to depend on this package and still build with Python 3.6. But if you want to test it out, let me know if it works for various versions of Python.

File: var/spack/repos/builtin/packages/py-entrypoints/package.py [hunk 3]
```diff
@@ -25,15 +25,22 @@
...
 
     version('0.2.2', '73bd7ce92c19b25dc5a20aff41be996a', expand=False)
 
+    depends_on('python@2.7:', type=('build', 'run'))
+
     depends_on('py-pip', type='build')
+    depends_on('py-configparser', when='^python@:2.8', type=('build', 'run'))
```
Comment (healther) [inline var/spack/repos/builtin/packages/py-entrypoints/package.py]: why does this work in your case but produced a infinite recursion in mine?

Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-entrypoints/package.py]: Because:
```python
depends_on('py-configparser@3.5:', when='^python@2.7', type=('build', 'run'))
```
means it depends on `py-configparser` only when building with Python 2.7. _Exactly_ 2.7. Since the default is 2.7.13 (not 2.7), it doesn't depend on `py-configparser`. Normally this would be fine and the dependency would be ignored, but the concretizer is currently broken (see #2793) resulting in the bug you see. At least, that's my understanding of the problem. @tgamblin knows more about the concretizer than me.

Keep in mind that none of these bugs are by design, and we would like to see them fixed just as much as you. @tgamblin is working on a complete overhaul of the concretizer, so I expect to see this bug solved in the next release.

File: var/spack/repos/builtin/packages/py-configparser/package.py (hunk 1 above)
Comment (healther) [inline var/spack/repos/builtin/packages/py-configparser/package.py]: ah I see, but just to be sure I understood correctly: A package that has a ```depends_on('py-configparser'``` and tries to build with ```python@3.6``` is technically pulling in to many dependencies, isn't it?

Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-configparser/package.py]: To be honest, this is such an intricate bug that I don't even remember anymore. It might work, it might not. From my experience, if you don't specify a version of Python (use the default) and the dependency isn't added for that version of Python, things crash. So it might work. But I can't remember what happens if you change the default in `packages.yaml`. Long story short: Python support is pretty broken for backports, and for Python 3.

**Comment by adamjstewart** [comment]
@lee218llnl Can you comment on this PR and make sure I'm not doing anything stupid? @svenevs I vaguely recall you having similar problems with `configparser` or `ordereddict`. Did I fix anything (or make it worse)?

File: var/spack/repos/builtin/packages/py-entrypoints/package.py (hunk 3 above)
Comment (healther) [inline var/spack/repos/builtin/packages/py-entrypoints/package.py]: Ah okay, so again just checking my understanding ```when='^python@2.7'``` should have worked, if not for the concretizer bug? And ```when='^python@2.7.0:2.7.99'``` should have worked in spite of it?

**Comment by healther** [comment]
Thanks for the quick help, that fixes my problems, if you want me to test anything, feel free to ask. Otherwise this closes #4327 from my side

File: var/spack/repos/builtin/packages/py-entrypoints/package.py (hunk 3 above)
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-entrypoints/package.py]: There's multiple concretizer bugs. The fact that `@2.7` doesn't work could be considered either a 'feature' or a 'bug'. But yes, `@2.7.0:2.7.99` should work either way. `@2.6.0:2.6.99` doesn't work due to a different concretizer bug.

**Comment by lee218llnl** [comment]
This looks good to me, but I don't think the import_modules tests are run. I don't see any output in build.out and I can change the module name to some garbage that clearly wouldn't import and the installation still succeeds.

**Comment by adamjstewart** [comment]
The `import_modules` tests are only run when you ask for them:
```
$ spack install --run-tests py-entrypoints
```

**Comment by lee218llnl** [comment]
@adamjstewart here's what I see (note I changed the import_modules to "entrypoints2" to try and get the install with tests to fail:

```
[lee218@rztrona2:spack.pr4330]$ git branch
  develop
* fixes/entrypoints
[lee218@rztrona2:spack.pr4330]$ grep import_modules var/spack/repos/builtin/packages/py-entrypoints/package.py
    import_modules = ['entrypoints2']
[lee218@rztrona2:spack.pr4330]$ ./bin/spack -v install --run-tests py-entrypoints
==> python is already installed in /nfs/tmp2/lee218/delete/spack.pr4330/opt/spack/linux-rhel7-x86_64/gcc-4.8.5/python-2.7.13-5xsz3tmkpq3xyuklugxh2xcbiqtsvgqd
==> py-pip is already installed in /nfs/tmp2/lee218/delete/spack.pr4330/opt/spack/linux-rhel7-x86_64/gcc-4.8.5/py-pip-9.0.1-tpkmudcnebegxbrbdl24rmxjhnytjom4
==> py-configparser is already installed in /nfs/tmp2/lee218/delete/spack.pr4330/opt/spack/linux-rhel7-x86_64/gcc-4.8.5/py-configparser-3.5.0-qteumbprmqn4npoifaikl6cqihr6q776
==> Installing py-entrypoints
==> Using cached archive: /nfs/tmp2/lee218/delete/spack.pr4330/var/spack/cache/py-entrypoints/py-entrypoints-0.2.2.None
==> Already staged py-entrypoints-0.2.2-ab73hobrlfh6t24anjhs7f2ajrbn2jq4 in /nfs/tmp2/lee218/delete/spack.pr4330/var/spack/stage/py-entrypoints-0.2.2-ab73hobrlfh6t24anjhs7f2ajrbn2jq4
==> Ran patch() for py-entrypoints
==> Building py-entrypoints [PythonPackage]
==> Executing phase : 'install'
==> '/nfs/tmp2/lee218/delete/spack.pr4330/opt/spack/linux-rhel7-x86_64/gcc-4.8.5/py-pip-9.0.1-tpkmudcnebegxbrbdl24rmxjhnytjom4/bin/pip' 'install' '/nfs/tmp2/lee218/delete/spack.pr4330/var/spack/stage/py-entrypoints-0.2.2-ab73hobrlfh6t24anjhs7f2ajrbn2jq4/entrypoints-0.2.2-py2.py3-none-any.whl' '--prefix=/nfs/tmp2/lee218/delete/spack.pr4330/opt/spack/linux-rhel7-x86_64/gcc-4.8.5/py-entrypoints-0.2.2-ab73hobrlfh6t24anjhs7f2ajrbn2jq4'
Processing /nfs/tmp2/lee218/delete/spack.pr4330/var/spack/stage/py-entrypoints-0.2.2-ab73hobrlfh6t24anjhs7f2ajrbn2jq4/entrypoints-0.2.2-py2.py3-none-any.whl
Requirement already satisfied: configparser>=3.5; python_version == "2.7" in /nfs/tmp2/lee218/delete/spack.pr4330/opt/spack/linux-rhel7-x86_64/gcc-4.8.5/py-configparser-3.5.0-qteumbprmqn4npoifaikl6cqihr6q776/lib/python2.7/site-packages (from entrypoints==0.2.2)
Installing collected packages: entrypoints
Successfully installed entrypoints-0.2.2
==> Successfully installed py-entrypoints
  Fetch: 0.04s.  Build: 3.94s.  Total: 3.98s.
[+] /nfs/tmp2/lee218/delete/spack.pr4330/opt/spack/linux-rhel7-x86_64/gcc-4.8.5/py-entrypoints-0.2.2-ab73hobrlfh6t24anjhs7f2ajrbn2jq4
```

**Comment by adamjstewart** [comment]
@lee218llnl You won't see any of the output of the build if you aren't building with the verbose flag.

**Comment by lee218llnl** [comment]
@adamjstewart I did run spack -v install. Also, note that the test should failed since I "broke" the module name that the test should import. Am I missing something here?

**Comment by adamjstewart** [comment]
Oh wait, you're completely right, it works for other Python packages but not `py-entrypoints`. @alalazo Can you explain this? Is it because I override `phases` in the package? Adding:
```python
from spack.package import PackageBase

run_after('install')(PackageBase._run_default_install_time_test_callbacks)
```
to the package gets it to run the tests correctly. Although the tests fail because of:
```
ImportError: No module named backports
```
